ENIGMAsystem/SHELL/Universal_System/Extensions/ProcInfo/Win32/procinfo.cpp

```diff
@@ -0,0 +1,283 @@
+﻿/*
+
+ MIT License
+ 
+ Copyright © 2020 Samuel Venable
+ 
+ Permission is hereby granted, free of charge, to any person obtaining a copy
+ of this software and associated documentation files (the "Software"), to deal
+ in the Software without restriction, including without limitation the rights
+ to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ copies of the Software, and to permit persons to whom the Software is
+ furnished to do so, subject to the following conditions:
+ 
+ The above copyright notice and this permission notice shall be included in all
+ copies or substantial portions of the Software.
+ 
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ SOFTWARE.
+ 
+*/
+
+#include "../procinfo.h"
+#define byte __windows_byte_workaround
+#include <windows.h>
+#undef byte
+#include <tlhelp32.h>
+#include <process.h>
+#include <psapi.h>
+#include <cstddef>
+#include <vector>
+#include <cwchar>
+
+using std::string;
+using std::to_string;
+using std::wstring;
+using std::vector;
+using std::size_t;
+
+static inline wstring widen(string str) {
+  size_t wchar_count = str.size() + 1;
+  vector<wchar_t> buf(wchar_count);
+  return wstring{ buf.data(), (size_t)MultiByteToWideChar(CP_UTF8, 0, str.c_str(), -1, buf.data(), (int)wchar_count) };
+}
+
+static inline string narrow(wstring wstr) {
+  int nbytes = WideCharToMultiByte(CP_UTF8, 0, wstr.c_str(), (int)wstr.length(), NULL, 0, NULL, NULL);
+  vector<char> buf(nbytes);
+  return string{ buf.data(), (size_t)WideCharToMultiByte(CP_UTF8, 0, wstr.c_str(), (int)wstr.length(), buf.data(), nbytes, NULL, NULL) };

```

Comment:
Don't we have these in `shared/` yet? If not, the ones in the engine should be moved there and your extension should link them.

Reply:
The reason I don't do that currently is because I want these files to be self-contained, meaning they can build independently apart from the rest of the engine - this extension for enigma I am also building as a dll for a gamemaker extension, and this makes that easier to maintain. But if you guys don't care about making my life easier, that's ok, just confirm you still want me to change it and I'll go right ahead and do it. lel

Reply:
Well the way I see it is you are bordering between an official ENIGMA extension and a regular user extension. I am not so convinced this needs to be in the main ENIGMA, since none of it is actual GM API? I am also not completely convinced of the necessity for its use case either. Regardless, ideally, all main official default included ENIGMA extensions would share code in a modular way. If maintainability concerns you then, honestly, just distributing this on the forum or etc as an unofficial user extension would be appropriate.

Reply:
Well, ENIGMA by definition is more than a GameMaker clone, but rather, an augmentation, so it would make sense if we had stuff in our engine that GameMaker doesn't, which is why I thought this extension might be acceptable. I'll go ahead and move those functions to shared like you wanted in a different pr, and once that gets merged we merge those changes into this branch and I'll include the functions from the shared header.

Since the last time I've talked with Josh, you, him, and fundies seem to all be on the same page that this doesn't belong in the engine. I'll close this pr if I can confirm from Josh he still feels this way, and I won't be surprised if he hasn't, as nothing has changed that would influence his opinion, I'm just reluctant to seeing all of this hard work I put into it not being an official part of ENIGMA. I can't get everything I want in life, jaja...

Reply:
I definitely see that point of view, that ENIGMA certainly can have features that are not in GM, but that doesn't mean the whole world as you already see. We have to be careful about going too far overboard with features that aren't really going to be used that essentially, lest we make the engine too hard to maintain.

I wasn't necessarily decided on the usefulness of it though either, so I wasn't saying this isn't a useful or essential extension. Just see what Josh thinks about what I said.

Reply:
You guys are just being honest, and I can't fault you for that. Will do.